Add support for page calls and fix use of _last_visit

alisdair opened the pull request:
Page events are pushed to Customer.io as custom events, with the "type" field set to "page", and the "name" field set to the URL of the page.

alisdair committed:
Add support for page calls

Page events are pushed to Customer.io as custom events, with the "type"
field set to "page", and the "name" field set to the URL of the page.

alisdair committed:
Don't send _last_visit for identify or track

The expectation from Customer.io is that the `_last_visit` parameter
should only be sent on page requests, as it indicates when the user was
active on the site. Identify calls and event tracking should not update
this parameter.

This commit removes the calls to `visit` for identify and track, and
updates the tests to validate that identify, group, and track all only
send one request.

We still update the `_last_visit` parameter for page requests.

Fixes #6

alisdair commented:
Added a commit to fix #6.

philfreo commented:
Thanks for the quick PR @alisdair! 

@gjohnson Do you think #6 can get merged / rolled into Segment.com soon?

sperand-io commented on lib/index.js:
```diff
@@ -33,15 +33,13 @@ var CustomerIO = module.exports = integration('Customer.io')
-    .send(mapper.identify(identify))
-    .end(self.handle(fn));
-  });
+  if (!identify.active()) return setImmediate(fn);
```
i don't think we should return early here — `msg.active()` returns false in the case when a customer sends a call because they want it recorded (an offline event happened or updating a user profile) but they specifically don't want to update `last_visit` or have the call count towards a session in tools where we can opt out from that. I'm fine with removing `last_visit` altogether in `identify` and `track` if that aligns better with c.io expectations, but it doesn't seem necessary to skip identifies and tracks that signal "this call is not part of an active session". 

thoughts? :)

alisdair replied:
Ah, I see. I didn't understand the semantics of `active()` correctly.

I've pushed up a commit that I think does what you suggested. Please let me know if there are any other changes needed!

sperand-io commented:
Looks awesome. One comment inline! @alisdair

alisdair committed:
Don't early-return from identify/track if inactive

We should make calls to identify & track whether the call is part of an
active session or not. We only update `_last_visit` for page views and
other explicit calls to `visit`.

alisdair commented:
@sperand-io Addressed, thanks! Let me know if you want anything else done or need the commits squashed.

alisdair commented:
Hi @sperand-io, anything else I can do to move this forward?

guillaumecabane commented:
Hey @alisdair thanks for the commit. @sperand-io I'd like to weight in on this change, as this is currently a serious issue for us, as we send page calls to compensate the fact that we have a one-page webapp. As of now, Customer.io does not get those page calls, and since there is not url change or page reload, I don't get the data in Customer.io

Additionally, I would say we should do the same for the Screen events :)

sperand-io commented:
lgtm — sorry for the delay everyone. thank you so much for the great work @alisdair!

sperand-io commented:
@alisdair I would love to add support for screen calls as well — would that make sense? What is the typical naming pattern for app screens (in lieu of a URL like we use in page)?
